### PR TITLE
Heap queue stores only peers that were not added to p2p server

### PR DESCRIPTION
**Pull request by dshulyak**
The primary goal of this change is to keep a whitelist of peers
that are managed by the topic pool while also preventing the same peer
from being selected from heap queue multiple times.

Closes: https://github.com/status-im/status-go/issues/973

### CODE REVIEW

File: geth/peers/topicpool.go [hunk 1]
```diff
@@ -64,36 +69,46 @@ func (t *TopicPool) addToPeerPool(peer *peerInfo) {
...
 }
 
-func (t *TopicPool) removeFromPeerPool(nodeID discv5.NodeID) {
-	peer, ok := t.peerPool[nodeID]
-	if !ok {
-		return
-	}
-	delete(t.peerPool, nodeID)
-	heap.Remove(&t.peerPoolQueue, peer.index)
+// addToQueue cannot be called if peer is no the pool
```
Comment (themue) [inline geth/peers/topicpool.go]: Would like to see the intention of the function as comment here and limitations as second sentence or internal comment. 

What does it mean here "cannot be called"? Would expect an error then. Or do you mean that it will never happen?

File: geth/peers/topicpool_test.go [hunk 2]
```diff
@@ -182,10 +183,63 @@ func (s *TopicPoolSuite) TestTheMostRecentPeerIsSelected() {
...
+	peer3 := discv5.NewNode(discv5.NodeID{3}, s.peer.Self().IP, 32311, 32311)
+
+	s.topicPool.processFoundNode(s.peer, peer1)
+	s.topicPool.processFoundNode(s.peer, peer2)
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
+	s.topicPool.ConfirmDropped(s.peer, discover.NodeID(peer2.ID))
+	s.Len(s.topicPool.peerPool, 1)
+	s.Contains(s.topicPool.peerPool, peer2.ID)
+	s.topicPool.processFoundNode(s.peer, peer3)
```
Comment (themue) [inline geth/peers/topicpool_test.go]: Expecting `peer3` is dropped here too. Am I right?

**Review by themue — APPROVED** [review]
LGTM, only two minor comments.

File: geth/peers/topicpool_test.go [hunk 3]
```diff
@@ -182,10 +183,63 @@ func (s *TopicPoolSuite) TestTheMostRecentPeerIsSelected() {
 	s.topicPool.processFoundNode(s.peer, peer2)
 	s.topicPool.processFoundNode(s.peer, peer3)
 	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer3.ID))
```
Comment (adambabik) [inline geth/peers/topicpool_test.go:187]: Why do we need these lines? These peers are immediately dropped below.

File: geth/peers/topicpool_test.go (hunk 2 above)
Comment (adambabik) [reply geth/peers/topicpool_test.go]: No, it was just found by Discovery protocol but the TopicPool did not connect to it.

File: geth/peers/topicpool_test.go [hunk 4]
```diff
@@ -182,10 +183,63 @@ func (s *TopicPoolSuite) TestTheMostRecentPeerIsSelected() {
...
+	peer1 := discv5.NewNode(discv5.NodeID{1}, s.peer.Self().IP, 32311, 32311)
+	peer2 := discv5.NewNode(discv5.NodeID{2}, s.peer.Self().IP, 32311, 32311)
+	s.topicPool.processFoundNode(s.peer, peer1)
+	s.topicPool.processFoundNode(s.peer, peer2)
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer1.ID))
+	s.topicPool.ConfirmAdded(s.peer, discover.NodeID(peer2.ID))
+	s.topicPool.ConfirmDropped(s.peer, discover.NodeID(peer2.ID))
+	s.topicPool.ConfirmDropped(s.peer, discover.NodeID(peer1.ID))
+
+	s.Contains(s.topicPool.peerPool, peer2.ID)
```
Comment (adambabik) [inline geth/peers/topicpool_test.go]: Should we also check that `peer1` is not in the `peerPool`? One may wonder what happened to `peer1`.

**Review by adriacidre — APPROVED** [review]
looks good

**Review by adambabik — COMMENTED** [review]
LGTM

Can we also clarify the difference between `peerPool` and `peerPoolQueue`? After this change, they will not be in sync.

Maybe something like that:
```go
peerPool       map[discv5.NodeID]*peerInfoItem // found but not connected peers
```
to
`// contains found and requested to be connected peers but not confirmed`
and 
```go
peerPoolQueue  peerPriorityQueue               // priority queue to find the most recent peer
```
to 
`// priority queue to find the most recently discovered peers; does not containt peers requested to connect`

Also, what's gonna happen when we add a peer to `peerPool` and to the p2p server but that peer will never actually connect? From the code, it looks like it won't be ever removed from the `peerPool` (which should not be a problem) and if it gets discovered again, its `discoveredTime` will be updated and handled properly. Is that correct?

**Comment by dshulyak** [comment]
> Also, what's gonna happen when we add a peer to peerPool and to the p2p server but that peer will never actually connect? From the code, it looks like it won't be ever removed from the peerPool (which should not be a problem) and if it gets discovered again, its discoveredTime will be updated and handled properly. Is that correct?

yes, if it is never connected - it will be purged from p2p server only when app reloads or never
not that big of a problem, but in future we could think about some timeout for adding a peer to p2p server

File: geth/peers/topicpool_test.go (hunk 3 above)
Comment (dshulyak) [reply geth/peers/topicpool_test.go:187]: because removal logic will be executed only if peer is in the connectedPeers map, and we want it to be executed because confirmDropped adds peer to peer queue if peer was dismissed

File: geth/peers/topicpool.go (hunk 1 above)
Comment (dshulyak) [reply geth/peers/topicpool.go]: makes sense, better to add protection instead of such vague comment

**Review by divan — COMMENTED** [review]
I'm not sure I understand all its implications of this change. 
The main concern is that `peerPoolQueue` should serve as a helper to speed up `find latest added peer`  lookup (we could just iterate over `peerPool` comparing timestamps, right?), but it ends up being a separate collection of peers with its own logic. Moreover, this logic is codified in the data type itself (`index`/`notQueuedIndexz` I mean), which makes it really hard to follow and further reason about the logic. Ideally, I would love to keep this `add`/`pop`/`remove` peers logic  in one place (maybe separate type), and all the logic of whitelisting/caching/speeding up things to be abstracted away.

Plus, can we come up with better name than `peerPool`? We now have abstractions tree `PeerPool`->`[]TopicPool`->`peerPool`/`peerPoolQueue`, where `PeerPool` and `peerPool` are totally different types/abstractions.

**Comment by themue** [comment]
Thanks for your review @divan . Reminds me of not focus on the pure code too much bus also take a step back, trying to get a bigger picture, and possibly compare it to an own solution.

**Comment by dshulyak** [comment]
i agree that having `index` usage is confusing but i don't see a better way to handle re-ordering when discovery updates time

i will rename peer pool though to pendingPeers

i mentioned in the original issue that having heap queue might be suboptimal, as updating all known peers is better strategy, cause we have rebalancer that will disconnect peers above max limit. it will simplify codebase but thats not for this change

File: geth/peers/topicpool.go [hunk 5]
```diff
@@ -53,51 +58,63 @@ type TopicPool struct {
...
-	peerPoolQueue  peerPriorityQueue               // priority queue to find the most recent peer
-	connectedPeers map[discv5.NodeID]*peerInfo     // currently connected peers
+	pendingPeers         map[discv5.NodeID]*peerInfoItem // contains found and requested to be connected peers but not confirmed
+	discoveredPeersQueue peerPriorityQueue               // priority queue to find the most recently discovered peers; does not containt peers requested to connect
+	connectedPeers       map[discv5.NodeID]*peerInfo     // currently connected peers
 
 	cache *Cache
 }
 
 func (t *TopicPool) addToPeerPool(peer *peerInfo) {
```
Comment (adambabik) [inline geth/peers/topicpool.go]: Should we also change function names to `addPendingPeer()` etc?

**Review by divan — APPROVED** [review]
@dshulyak awesome. I guess refactoring can be done in a separate PR once we fine more optimal/cleaner solution.

**Review by adambabik — APPROVED** [review]
